### ISSUE TITLE
Segmentation fault (SIGSEGV) on all new connections

### ISSUE THREAD

**Issue by VisionaryMind** [GoogleCloudPlatform/cloudsql-proxy#749]
## Bug Description

This is possibly a duplicate of Issue #[543](https://github.com/GoogleCloudPlatform/cloudsql-proxy/issues/543) , but according to my understanding, that bug should have been resolved by now. Starting today, I am seeing SIGSEGV signals thrown for all new connections routed through the cloud_sql_proxy. It was working perfectly fine up until now, and nothing has changed in our configuration.

The proxy may be started without issue:

```
2021/04/21 09:52:06 Rlimits for file descriptors set to {Current = 8500, Max = 1048576}
2021/04/21 09:52:06 using credential file for authentication; email=account@xxxx.iam.gserviceaccount.com
2021/04/21 09:52:07 Listening on /cloudsql/xxxx:us-central1:xxxx-mysql for xxxx:us-central1:xxxx-mysql
2021/04/21 09:52:07 Ready for new connections
```

But as soon as a new connection is attempted, the segmentation violation occurs:

```
2021/04/21 09:52:39 New connection for "xxxx:us-central1:xxxx-mysql"
2021/04/21 09:52:39 refreshing ephemeral certificate for instance xxxx:us-central1:xxxx-mysql
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x95c50f]

goroutine 114 [running]:
github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy.(*Client).startRefresh.func1(0xc0000242a0, 0xc0001c6300, 0x7ffea02d78fe, 0x34, 0x45d964b800)
	/workspace/proxy/proxy/client.go:285 +0x38f
created by github.com/GoogleCloudPlatform/cloudsql-proxy/proxy/proxy.(*Client).startRefresh
	/workspace/proxy/proxy/client.go:264 +0x85
```

## Example code (or command)

I am running cloud proxy with the following command:
```
cloud_sql_proxy -dir=/cloudsql -instances=xxxx:us-central1:xxxx-mysql \
                  -credential_file=/keylocation/key.json
```
## Environment

1. OS type and version: Ubuntu 20.04 LTS
2. Cloud SQL Proxy version: 1.21.0+linux.amd64

This is a very big problem for us, as it has shutdown our entire analysis / dev pipeline. The issue is happening on all workstations, regardless of OS. What could be the problem?

**Comment by enocom** [comment]
This is a duplicate of #745.

**Comment by VisionaryMind** [comment]
> This is a duplicate of #745.

However, I am running the proxy with "-ip_address_types=PRIVATE".

**Comment by enocom** [comment]
I assume you mean **not** running with that flag. So I've left the issue open to investigate.

**Comment by VisionaryMind** [comment]
I attempted to reproduce the error using v1.20.2, and it produces a different error:

```
2021/04/21 10:16:50 Throttling refreshCfg(xxxx:us-central1:xxxx-mysql): it was only called 157.498944ms ago
2021/04/21 10:16:50 couldn't connect to "xxxx:us-central1:xxxx-mysql": googleapi: Error 409: The instance or operation is not in an appropriate state to handle the request., invalidState
```

**Comment by enocom** [comment]
That's the problem here: when the proxy receives an error from the back end when trying to refresh the certificate, the proxy panics and doesn't report the error. We'll cut a patch release today with the fix from #745. In the meantime, the error you see in v1.20.2 points to issue with your database instance. Have you connected to the instance successfully in the past?

**Comment by enocom** [comment]
We ended up cutting a new minor version: [v1.22.0](https://github.com/GoogleCloudPlatform/cloudsql-proxy/releases/tag/v1.22.0).
